std/math/emulated/hints.go

```diff
@@ -188,17 +188,35 @@ func subPaddingHint(mod *big.Int, inputs, outputs []*big.Int) error {
 }
 
 func (f *Field[T]) computeSubPaddingHint(overflow uint, nbLimbs uint, modulus *Element[T]) *Element[T] {
+	// we compute the subtraction padding hint in-circuit. The padding has satisfy:
+	// 1. padding % modulus = 0
+	// 2. padding[i] >= (1 << (bits+overflow))
+	// 3. padding[i] + a[i] < native_field for all valid a[i] (defined by overflow)
 	var fp T
 	inputs := []frontend.Variable{fp.NbLimbs(), fp.BitsPerLimb(), overflow, nbLimbs}
 	inputs = append(inputs, modulus.Limbs...)
+	// compute the actual padding value
 	res, err := f.api.NewHint(subPaddingHint, int(nbLimbs), inputs...)
 	if err != nil {
 		panic(fmt.Sprintf("sub padding hint: %v", err))
 	}
+	maxLimb := new(big.Int).Lsh(big.NewInt(1), fp.BitsPerLimb()+overflow)
+	maxLimb.Sub(maxLimb, big.NewInt(1))
 	for i := range res {
+		// ensure that condition 3 always holds. As we compute the padding hint
+		// depending on the overflow of the inputs, then can use the maximum
+		// possible value
 		f.checker.Check(res[i], int(fp.BitsPerLimb()+overflow+1))
```

Comment:
@ivokub  so, I read, using range checks, ensure:
1. res[i] < ( 1 << (bits_with_overflow + 1) )
2. res[i] -  ( 1 << bits_with_overflow) + 1 < ( 1 << (bits_with_overflow + 1) )

Doesn't 2. suffice (i.e. isn't 1 a sub constraint of 2?)

Reply:
I think you are right, I'll fix it.

Reply:
there might be some edge cases with very small emulated and scalar field and or very large, but, didn't find any since I assume overflow is correctly computed and won't allow for bogus values (don't take my word for it just spent couple minutes on it)

Reply:
So there are two things what we want to check:
1. that the width of the padding limb is not bigger than `nbBits+overflow+1` (i.e. `padding < 2^{nbBits+overflow+1}`) -- we need to ensure this so that when we add the padding to `a[i]` that we wouldn't overflow the native field.
2. that the padding limb is not smaller than maximum permitted value for non-native element limb i.e. `2^{nbBits+overflow}-1` -- this is to ensure that we wouldn't underflow the circuit when subtracting `b[i]` from padded `a[i] + s[i]`

The first range check checks condition 1 directly. For the second one we instead subtract the maximum permitted value and check that the difference is not big (it is big when we would have underflown).

But you are right, I think we can merge the two checks together

Reply:
Removed one range check and tried to clarify the comments.